Add a workflow for cache garbage collection

desrosj opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This adds a GitHub Actions workflow that performs garbage collection on GHA cache keys that are associated with a PR once it's closed out.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Once a PR is merged and/or closed, the cache keys created within associated workflow runs are useless.

Access restrictions provide cache isolation and security by creating a logical boundary between different branches or tags. Workflow runs can restore caches created in either the current branch or the default branch (usually main). If a workflow run is triggered for a pull request, it can also restore caches created in the base branch, including base branches of forked repositories.

When a cache is created by a workflow run triggered on a pull request, the cache is created for the merge ref (`refs/pull/.../merge`). Because of this, the cache will have a limited scope and can only be restored by re-runs of the pull request. It cannot be restored by the base branch or other pull requests targeting that base branch. This happens any time a PR changes lock files (which are used to create cache keys in most cases), or when a cache key for the primary branch is evicted or missing.

Each repository is allotted 10 GB worth of cache storage for GHA. Once that is exceeded, GitHub will proceed to evict cache entries until total storage falls below the allowed threshold. As far as I can tell, how GHA decides which cache keys to evict is not clear. By proactively evicting caches that we know are unlikely to be used again, we can ensure more important ones remain available longer.

You can read more about [caching within GitHub Actions and how this is managed in the documentation](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows).

desrosj committed:
Add a workflow for cache garbage collection.

desrosj committed:
Change to all `pull_request` types for testing purposes.

desrosj committed:
Remove extra `.`.

desrosj committed:
Fix syntax error.

dmsnell commented on .github/workflows/cleanup-caches.yml:
```diff
@@ -0,0 +1,43 @@
+name: Cache Garbage Collection
+
+on:
+  pull_request:
```
is this here for testing? later on maybe we run this on merge into trunk? or only when closing a PR?

desrosj replied:
Ya, sorry. I forgot to re-add the `types` filter for only running on `closed` after testing. All set now!

dmsnell commented on .github/workflows/cleanup-caches.yml:
```diff
@@ -0,0 +1,43 @@
+
+            // Find the caches related to this pull request.
+            for ( let i = 0; i < action_caches.data.actions_caches.length; i++ ) {
+              if ( '${{ github.ref }}' == action_caches.data.actions_caches[ i ].ref ) {
```
I think this is the `$GITHUB_REF` value isn't it? which points to the merge branch of the PR? I think this will change every time update a PR or the target branch (usually `trunk`) updates.

should we be matching based on something like the branch name, or the PR number?

desrosj replied:
According to the [documentation](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context), `github.ref` context will always be `refs/pull/<pr_number>/merge`, which matches the pattern defined in the caching documentation linked in the PR description. I'll drop some information about how I tested this in general directly on the PR, but I confirmed this in my testing.

dmsnell replied:
> which matches the pattern defined in the caching documentation

I'm not sure I understand what you are suggesting. I am thinking that a given PR could create "15 caches" and this script is deleting "1 cache" because its only looking at the cache of the merge branch _when the PR merged_, but those merge branches update on two events:

 - the PR's branch is changed in any way
 - the target branch is changed in any way

so every commit, every rebase, every other merged PR, updates these merge branches, which I am guessing means some fraction of them create new cache entires.

dmsnell reviewed (commented):
I'm sad we need a cleanup task, but glad we're getting one 😄

kevin940726 commented:
Thank you for starting this! However, I'm not sure if we even need this. Based on the doc you linked in the PR:

> If you exceed the limit, GitHub will save the new cache but will begin evicting caches until the total size is less than the repository limit.

I think it's basically an LRU cache. As long as the primary cache is used frequently (which is often the latest cache on trunk, the most common branch to be based on), it should rarely be evicted. On the other hand, a cache created by a closed PR will be evicted once it's no longer used by anyone and there's a new cache in town.

What this workflow does is proactively remove potentially unused caches. While this looks harmless, I'm not so sure about the benefits it brings. I think we can still try this, but we might need a way to collect numbers of the impact.

desrosj committed:
Re-add the `pull_request` type.

desrosj commented:
This is definitely somewhat experimental. But I think it's worth trying.

I reread the documentation page linked in the PR, and [hidden within the last section](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#force-deleting-cache-entries) is a more clear answer to how GHA goes about cleaning up cache keys.

> Once a repository has reached its maximum cache storage, the cache eviction policy will create space by deleting the oldest caches in the repository.

That means caches created in `trunk` can be deleted if the key does not change and enough PRs are opened/time passes. My thinking is that if we proactively purge cache keys that we know are no longer needed, keys we know are beneficial have a higher likelihood of sticking around.

With this in mind, we could also add another step that detects when a change to a lock file occurs and deletes any caches with the old file hash in the key. I think that's a little much for this first iteration.

> I'm sad we need a cleanup task, but glad we're getting one 😄

I don't necessarily think we _need_ a workflow for this as GitHub does clean up caches on its own. But it's likely by implementing our own cleanup, we can cache longer and more effectively. I do wish they allowed us to configure priority for cleanup, though. Something like "closed PRs first, non-main branch next, etc." instead of just "oldest key first".

I also wanted to add some testing details as I forgot to include those. To test, I waited until there were no workflows running and found a cache key associated with `trunk`. I deleted it, and re-ran a workflow on this PR that utilized that key. When the workflow finished, a new key was added scoped to this PR. Then I removed the `types` filter on this PR and pushed. That triggered the workflow and I confirmed the cache was deleted in the repository's GHA Cache management page.

kevin940726 commented:
> Once a repository has reached its maximum cache storage, the cache eviction policy will create space by deleting the oldest caches in the repository.

I read this as "oldest" meaning the least recently used cache. It's the last updated cache but not the last created cache, I could be wrong though 😅 . IIUC, this means we shouldn't need to worry about trunk's cache as it's most frequently used.

desrosj commented:
I'm going to close this out since it's been almost a year. It's possible this may still be worth exploring, but our workflows and the GHA service has changed a bunch, so we'll need to evaluate from the start again.
